### PR TITLE
Marker Maze classification

### PR DESCRIPTION
**Pull request by markmellors**
tested and working.  runs noticably faster than non-threaded version (parameters will need to be retuned)

### CODE REVIEW

File: create_aruco_dict.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+#small_dict = aruco.generateCustomDictionary(6, 3)
+NUM_OF_MARKERS = 6
+small_dict = aruco.Dictionary_create(NUM_OF_MARKERS, 3)
+'''
+    drawMarker(...)
+        drawMarker(dictionary, id, sidePixels[, img[, borderBits]]) -> img
+'''
+             
+print(small_dict)
```
Comment (robberwick) [inline create_aruco_dict.py]: switch to `logging.<LOG_LEVEL>`? Also, why is this here?

File: create_aruco_dict.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import sys
+sys.path.append('/usr/local/lib/python2.7/site-packages')
+import cv2
+import numpy as np
+import cv2.aruco as aruco
+
+#small_dict = aruco.generateCustomDictionary(6, 3)
```
Comment (robberwick) [inline create_aruco_dict.py]: Is this dead code?

File: create_aruco_dict.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+NUM_OF_MARKERS = 6
+small_dict = aruco.Dictionary_create(NUM_OF_MARKERS, 3)
```
Comment (robberwick) [inline create_aruco_dict.py]: I'm confused by this - is this supposed to be run as a standalone script? Should it not be defined as a function, with the multi line comment acting as a doc string?

File: marker_maze.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
+from img_base_class import *
+import cv2.aruco as aruco
+
+# Image stream processing thread
+class StreamProcessor(threading.Thread):
```
Comment (robberwick) [inline marker_maze.py:5]: This is somewhat common with the stuff in `rainbow.py`, isn't it? Perhaps we can refactor to have a base class for the `StreamProcessor`.. 

this is fine, but we should at least raise an issue to address the code duplication.

File: marker_maze.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+        self.CROP_WIDTH = 320
+        self.i = 0
+        self.TIMEOUT = 30.0
+        self.START_TIME = time.clock()
+        self.END_TIME = self.START_TIME + self.TIMEOUT
+        self.found = False
+        self.turn_number = 0
+        self.TURN_TARGET = 5
+        self.TURN_WIDTH = [30, 35, 35, 30, 35, 35]
+        self.NINTY_TURN = 0.8  #0.8 works if going slowly
```
Comment (robberwick) [inline marker_maze.py]: "NINTY"? :-)

File: marker_maze.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+    def process_image(self, image, screen):
+        screen = pygame.display.get_surface()
+        if self.turn_number >= self.TURN_TARGET:
+           logger.info("finished!")
+           self.finished = True
+        frame = image[30:190, (self.image_centre_x - self.CROP_WIDTH/2):(self.image_centre_x + self.CROP_WIDTH/2)]
+        # Our operations on the frame come here
+        gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
+        parameters =  aruco.DetectorParameters_create()
+        #print(parameters)
```
Comment (robberwick) [inline marker_maze.py]: Remove code if dead.

File: marker_maze.py [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+        '''    detectMarkers(...)
```
Comment (robberwick) [inline marker_maze.py]: Why the inline docstring?

File: marker_maze.py [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+            detectMarkers(image, dictionary[, corners[, ids[, parameters[, rejectedI
+            mgPoints]]]]) -> corners, ids, rejectedImgPoints
+        '''
+        #lists of ids and the corners beloning to each id
+        corners, ids, rejectedImgPoints = aruco.detectMarkers(gray, self.small_dict, parameters=parameters)
+        if ids != None:
+            #print ("found marker %s" % ids)
```
Comment (robberwick) [inline marker_maze.py]: ☠️ ?

File: marker_maze.py [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+                    #if there was a real error last time then do some damping
+                    turn -= self.TURN_D *(self.last_t_error - self.t_error)
+                turn = min(max(turn,-self.MAX_TURN_SPEED), self.MAX_TURN_SPEED)
+                #if we're rate limiting the turn, go slow
+                if abs(turn) == self.MAX_TURN_SPEED:
+                    self.drive.move (turn, self.STRAIGHT_SPEED/3)
+                else:
+                    self.drive.move (turn, self.STRAIGHT_SPEED)
+                self.last_t_error = self.t_error
+                #print(camera.exposure_speed)
```
Comment (robberwick) [inline marker_maze.py]: 💀 ?

File: marker_maze.py [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+                    self.turn_left()
+            self.found = False
+            self.last_t_error = 0
+        # Display the resulting frame
+        frame = pygame.surfarray.make_surface(cv2.flip(frame,1))
+        screen.fill([0,0,0])
+        screen.blit(frame, (0,0))
+        pygame.display.update()
+        if self.found:
+         img_name = str(self.i) + "Fimg.jpg"
```
Comment (robberwick) [inline marker_maze.py]: this manual `str()` and concatenation lark is unnecessary - Use string formatting to be more concise (and less error prone):
```
found_identifier = "F" if self.found else "NF"
img_name = "%d%simg.jpg" % (self.i, found_identifier)
```

**Review by robberwick — CHANGES_REQUESTED** [review]
Some minor changes neded, and we obviously still need to refactor out the `StreamProcessor` base class.

Was the `straight_line.py` meant to be included in this PR?

File: create_aruco_dict.py (hunk 2 above)
Comment (markmellors) [reply create_aruco_dict.py]: yes. I forgot this file would be included and hadn't tidied it up.

File: create_aruco_dict.py (hunk 3 above)
Comment (markmellors) [reply create_aruco_dict.py]: Sorry this isn't clearer. This is a standalone script to generate the image files for the Aruco markers we're using, so only needs to be run once.  that docstring should be converted into a comment, I think its purpose at the time was to  remind me of the arguments of drawMarker.

File: create_aruco_dict.py (hunk 1 above)
Comment (markmellors) [reply create_aruco_dict.py]: its mostly legacy from debugging, as with the later print. I don't think logging's needed here, since it should just run once

File: marker_maze.py (hunk 4 above)
Comment (markmellors) [reply marker_maze.py:5]: yes, this is as we discussed. issue raised

File: marker_maze.py (hunk 7 above)
Comment (markmellors) [reply marker_maze.py]: as from create_aruco_dict.py, it can be removed

File: marker_maze.py (hunk 8 above)
Comment (markmellors) [reply marker_maze.py]: dead

File: marker_maze.py (hunk 9 above)
Comment (markmellors) [reply marker_maze.py]: dead. side note: picamera's exposure_speed is misnamed. it should be exposure time,  as its time in microseconds (large is longer), not fractions of a second like the photography norm (large is shorter). I keep forgetting. I will add an actual note

File: marker_maze.py (hunk 10 above)
Comment (markmellors) [reply marker_maze.py]: good, thanks. you missed (or haven't yet got to) the code repetition I was expecting you to call me on. there's a nested if, and if either is false, I'd like the same thing to happen, at the moment that's achieved through duplication...

**Review by robberwick — APPROVED** [review]
